### ISSUE TITLE
Server side performance issue - useQuery, InMemoryCache

### ISSUE THREAD

**Issue by ITSWYoo** [apollographql/apollo-client#8638]
### 1. ssr `useQuery - fetchPolicy: cache-only`
  * When using `useQuery`, there is no need to track data changes on the server side. So we used `apolloClient.cache.readQuery` instead of `useQuery,` and performance improved.
    * Using `useQuery` reduces performance because of the logic to detect data changes, But I don't think it's necessary for 'SSR'.
    * If it is `ssr` in useQuery and `cache-only`, I think it is better if the internal logic is implemented to use the `readQuery` function.
    * so, we made a wrapping useQuery
 ```js
const useCacheOnlyQuery = <TData = any, TVariables = OperationVariables>(
  query: DocumentNode | TypedDocumentNode<TData, TVariables>,
  options?: QueryHookOptions<TData, TVariables>,
): { data: TData; loading: boolean } => {
  if (isServer) {
    const apolloClient = useApolloClient()

    if (options?.ssr === false) {
      return { data: null, loading: true }
    }

    // ssr
    const data = apolloClient.cache.readQuery({ query, variables: options?.variables, canonizeResults: false })

    return { data, loading: false }
  }

  // csr
  const { data, loading } = useQuery(query, {...options, fetchPolicy: 'cache-only'})

  return {
    data,
    loading,
  }
}
```
### 2. `InMemoryCache`
  * In server side rendering, diff and broadcastWatches functions of InmemoryCache affect the server cpu.
  * there is no need to track data changes on the server side.
  * so, we made a wrapping InMemoryCache 
```js
// inMemoryCacheForServerSide.ts
import { InMemoryCache } from '@apollo/client'
import { Cache } from '@apollo/client/cache/core/types/Cache'

// disable diff && broadcastWatches
export class InMemoryCacheForServerSide extends InMemoryCache {
  diff<T>(options: Cache.DiffOptions): Cache.DiffResult<T> {
    // Do nothing for performance
    return {}
  }

  broadcastWatches(): void {
    // Do nothing for performance
  }
}

// apolloClient.ts
const createApolloClient = (options: ApolloClientOption) => {

// ...

const cache = isServer
    ? new InMemoryCacheForServerSide({
        ...possibleTypes,
        typePolicies,
        resultCaching: false,
      })
    : new InMemoryCache({
        ...possibleTypes,
        typePolicies,
      })

return  new ApolloClient({
    cache,
    link: from([errorLink, persistedQueryLink, httpLink]),
    ssrMode: isServer,
    connectToDevTools: !isServer,
  })
}
```
* What are your thoughts on the above two points? 😄

**Comment by y-a-v-a** [comment]
We chose to do something similar but then by using `resultCaching: typeof window !== 'undefined',` to differentiate between server side and client side. When setting `resultCaching` to `true`, the memory usage runs completely out of hand due to deep nested `trie` ([this code](https://github.com/benjamn/wryware/blob/main/packages/trie/src/trie.ts)) objects. This especially was an issue when one has query results of several levels deep containing lists of several tens of items.

**Comment by phryneas** [comment]
@y-a-v-a the memory leaks you've been seeing should have been solved in Apollo Client 3.9, which has been released last week - see [this blog post](https://www.apollographql.com/blog/apollo-3-9-beta-feature-spotlight-the-memory-story).

@ITSWYoo Especially the second issue you are bringing up here looks like you are creating one Apollo Client instance and sharing that over multiple requests. 
Generally, when using Apollo Client on the server, you should create a new instance of Apollo Client for each incoming request.

As this is a very old issue and there's nothing really actionable (the SSR story is changing a lot right now and will not include `useQuery` going to the future with suspense) for us right now, I'm going to close this for now.

If you believe that there is something we should be acting on, please open a new issue.

**Comment by y-a-v-a** [comment]
@phryneas great news! I hope to find some time in the near future to upgrade! Thanks for letting me know.
